Clinton/der model

clinton0313 opened the pull request:
Model is mostly correct, but I am not a hundred percent certain. Couple points:

1) There are methods I have implemented that are consistent with the methods in the original paper's published github code (in particular the DeepEvidentialCallback), but these methods are suspiciously not mentioned in the paper: https://github.com/aamini/evidential-deep-learning

2) I have tried to reproduce the exact result of the cubic test as in the above github, but even while setting the same seeds I cannot get it exactly, and some of the variance around the training points are not as smooth so there is some uncertainty whether this is correct. Caveat is that I am able to produce many results that are similar in some way, but from seed to seed or initialization to initialization the results can vary wildly. A little put off though, since I have cloned amini's repo and realized that the perfect figure that he reports is only perfect for seed(1234) and if you change the seeds for his code, you will also get wildly different results of the model. Seems like he cherry picked his results. 

3) One avenue to potentially explore afterwards is maybe adding dropout and batch_normalization between layers to see if this can help regularize the results? 

4) There are two extra files that will be in the future deleted or moved out of the models folder: amini.pkl and model_test.py which is what I have been using to diagnose the model. (see final commit note).

clinton0313 committed:
Added loss functions for Deep Evidential Regression

clinton0313 committed:
Added tests for deep evidential regression loss functions

clinton0313 committed:
Script for testing loss function

clinton0313 committed:
Implemented DeepEvidentialNetwork. To be tested.

clinton0313 committed:
Completed and tested Deep Evidential network Architecture

clinton0313 committed:
Added builder function. Untested

clinton0313 committed:
Deep Evidential Network builder tests implemented

clinton0313 committed:
Amended architecture output to return well-behaved outputs evidential outputs and implemented corresponding unit test

clinton0313 committed:
Implemented alternative evidence activation functions for evidential parameters and corresponding test.

clinton0313 committed:
Merge branch 'clinton/der_architecture' into clinton/der_model

clinton0313 committed:
Merge branch 'clinton/der_loss_fn' into clinton/der_model

clinton0313 committed:
Removed loss script

clinton0313 committed:
Amended DER builder function with evidence_activation argument and amended corresponding unit test

clinton0313 committed:
Merge branch 'clinton/der_architecture' into clinton/der_model

clinton0313 committed:
Added Deep evidential regression model basic methods. Untested.

clinton0313 committed:
Debugged Deep Evidential Model

clinton0313 committed:
Added a build loss function and corresponding tests. Added some model tests.

clinton0313 committed:
Amended DER loss function builder to work in tensorflow training.

Fixed DER model loss tests.

clinton0313 committed:
Added learning rate model_test for DER

clinton0313 committed:
Marked deep evidential util tests.

clinton0313 committed:
Split up sample function in DER model and added appropriate test

clinton0313 committed:
Added a DeepEvidential Trajectory Sampler, untested. Added interface to facilitate inheritance and typing for it.

clinton0313 committed:
Added some of the model docstrings and placeholders for future model tests for DER.

clinton0313 committed:
Added a model sample test and fixed model sampling

clinton0313 committed:
Fixed loss functions and loss tests to be consistent with 2020 paper for DER

clinton0313 committed:
Fixed model and architecture tests for new loss functions and naming convention. Added dtype and scalar test to architecture test for DER

clinton0313 committed:
Added paper's cubic test to model_test. Not estimating OOD variance properly

clinton0313 committed:
Fixed crucial mistake in DER loss function. Predictions still seem unstable, possible square root missing in sigmas.

clinton0313 committed:
Amended incorrect sigma variances in sample to sigma **0.5, model is potentially working, but results seem unstable

clinton0313 committed:
models.py: Re-implemented DER  model to be consistent with amini's paper's code. adding:
~ multi-loss output
~ dual output
~ callback to update regulurazer weight
~ parameters for reg_weight, maxi_rate, epsilon
~ adjusted all other methods to account for above changes.

test_models.py: Adjusted and added tests for model changes:

~added test for dtype between reg_weight and input, outputs
~added tests to check for default callbacks
~added test to check for hardcoded DER loss function

utils.py:
~Adjusted loss functions to reduce_mean separately for DER
~Added DeepEvidentialCallback for diagnosis and updated regularizer weight

test_utils.py:
~Adjusted tests for new loss functions

architectures.py:
~Updated self._evidence with small float to keep evidence above 0

test_architecures.py:
~ Updated tests for changes in loss functions

test_predictions.py:
~ Implemented integration test for DER

extra files: to be deleted when testing complete
amini_cubic.pkl: exact cubic dataset from amini paper
model_test.py: Testing of cubic dataset cleaned up and can be ported into deep_evidential notebook in the future

clinton0313 committed:
Moved test files to notebooks and outputted numerous figures for different architectures.

ampudia19 commented on trieste/models/interfaces.py:
```diff
@@ -300,6 +300,31 @@ def sample_ensemble(self, query_points: TensorType, num_samples: int) -> TensorT
+    This is an interface for models that have a normal prior distribution over the outputs 
+    defined by a Normal Inverse Gamma higher order evidential distribution. these models can act
+    as probabilistic models by deriving estimates of epistemic uncertainty from the variance of mu
+    and aleatoric uncertainty from the expectation of sigma if it has a higher order distribution. 
```
Assuming the prior normal, we will always have at least two moments, which potentially makes the last clarification comment a bit redundant and confusing.

ampudia19 commented on trieste/models/keras/architectures.py:
```diff
@@ -233,3 +235,113 @@ def connect_layers(self) -> tuple[tf.Tensor, tf.Tensor]:
+
+        return output_layer
+
+    def _get_evidence(self, evidence: tf.Tensor) -> tf.Tensor:
```
This of course works, but I don't quite fully see the point of making _get_evidence be Tensor-specific, and not an array of three tensors where you then apply the transformations as in below.

clinton0313 replied:
Just seemed cleaner than split, apply, slice, +1, concat, return

ampudia19 commented on trieste/models/keras/builders.py:
```diff
@@ -80,3 +80,38 @@ def build_vanilla_keras_ensemble(
+    data: Dataset,
+    num_hidden_layers: int = 3,
+    units: int = 100,
+    activation: Union[str, tf.keras.layers.Activation] = "relu",
```
I'm guessing you have access to the cloned repo, but I do wonder whether they use softplus for all layers, considering it approximates the ReLu.

clinton0313 replied:
No he does not. He uses a relu. I once accidentally put in softplus for the activation and the results get a little crazy.

ampudia19 replied:
Indeed, while testing it the results were all over the place. Seems like a bit of a theme with DER.

ampudia19 commented on trieste/models/keras/utils.py:
```diff
@@ -98,3 +99,109 @@ def negative_log_likelihood(
+
+    omega = 2 * beta * (1 + v)
+
+    negative_log_likelihood = (
```
There's a naming conflict here, deep ensembles uses the negative_log_likelihood function. Not an issue, but perhaps we should avoid any naming overlap.

ampudia19 commented on trieste/models/keras/architectures.py:
```diff
@@ -25,6 +25,8 @@
 import tensorflow as tf
 import tensorflow_probability as tfp
 
+from trieste.models.keras.utils import normal_inverse_gamma_negative_log_likelihood, normal_inverse_gamma_regularizer
```
These are unused as you correctly compile the model in models.

ampudia19 commented on trieste/models/keras/utils.py:
```diff
@@ -98,3 +99,109 @@ def negative_log_likelihood(
+    gamma, v, alpha, _ = tf.split(y_pred, 4, axis=-1)
+    return tf.reduce_mean(tf.abs(y_true - gamma) * (2*v + alpha), axis=0)
+
+def deep_evidential_regression_loss(
```
This one is useless now, right? You use both losses in a list when compiling, and use the callback to calibrate lambda. To the best of my ability, I can't think nor see what use you give this function now.

ampudia19 commented on trieste/models/keras/utils.py:
```diff
@@ -98,3 +99,109 @@ def negative_log_likelihood(
+        self.epsilon = epsilon
+        self.verbose = verbose
+
+    def on_batch_end(self, epoch, logs=None):
```
on_batch_end has no epoch argument, and in any case you don't use it.

ampudia19 commented on trieste/models/keras/utils.py:
```diff
@@ -98,3 +99,109 @@ def negative_log_likelihood(
+        self.verbose = verbose
+
+    def on_batch_end(self, epoch, logs=None):
+        self.reg_weight.assign_add(self.maxi_rate * (logs["output_2_loss"] - self.epsilon))
```
I can't find this anywhere in the paper, but I recall you saying this comes from his paper code. Are maxi_rates and epsilons calibrated somehow, or are these tunable hyperparams?

clinton0313 replied:
It is not in the paper just in his code. They seem to just be hyperparameters. 

Here is the code for the paper's trainer: 
https://github.com/aamini/evidential-deep-learning/blob/main/neurips2020/trainers/evidential.py

The defaults are lambda = 0 (reg_weight), maxi_rate = 1e-4, epsilon = 1e-2. 

Here for the cubic tests he is using lambda = 1e-2 (reg_weight) and maxi_rate = 0:
https://github.com/aamini/evidential-deep-learning/blob/main/neurips2020/run_cubic_tests.py

Here he is instantiating the trainer with its defaults (line 47) to run the benchmark datasets:
https://github.com/aamini/evidential-deep-learning/blob/main/neurips2020/run_uci_dataset_tests.py

ampudia19 replied:
Have you tried saving the best model with a callback, like he does?

ampudia19 commented on trieste/models/keras/sampler.py:
```diff
@@ -106,3 +109,117 @@ def resample(self) -> None:
+            self._model.model.flattened_output_shape,
+            tf.shape(x)[-1],
+            message=f"""
+            IS THIS EVEN THE RIGHT TEST??
```
In the newer samplers they only check that the batch dimension is correct, a bit tautological since we set it so two lines above.

ampudia19 commented on trieste/models/keras/sampler.py:
```diff
@@ -106,3 +109,117 @@ def resample(self) -> None:
+            """
+        )
+
+        dist = tfp.distributions.Normal(self._mu, self._sigma)
```
These are not deterministic, but the solution here is straightforward and potentially compatible with graph execution (to use seeds in the tfp.distributions.Normal). I'll happily work on this once code is merged.

clinton0313 replied:
I pretty much just stopped working on this as soon as I realized I didn't understand the way the dimensions were working and waited for Hrvoje's response to my questions. Haven't touched the sampler since. Still a bit confusing to me at first thought whether it is the prediction which would just be mu or the distribution that constitutes the trajectory

ampudia19 replied:
In the case of ensembles and you can do either (notice he has this "use_samples" argument to return the distribution and sample from it, and otherwise the default behavior is the predictions from networks in the ensemble). Since I'm also stuck at the sampler, let's just see what Hrvoje has to say, and (if you don't mind) I'll take care of it.

ampudia19 commented on trieste/models/keras/sampler.py:
```diff
@@ -106,3 +109,117 @@ def resample(self) -> None:
+        tf.debugging.Assert(self._evidential_parameters is not None, self._evidential_parameters, \
+            "Trajectory sampler needs to have been called ot initialize before resampling.")
+        gamma, lamb, alpha, beta = tf.split(self._evidential_parameters, 4, axis = -1)
+        self._mu, self._sigma = self._model.sample_normal_parameters(gamma, lamb, alpha, beta, 1)
```
a possibly more computationally friendly way is to reset the seeds that you then use to draw from the tfp.distributions.Normal, as per my comment above.

ampudia19 commented on trieste/models/keras/models.py:
```diff
@@ -334,25 +343,159 @@ def optimize(self, dataset: Dataset) -> None:
+                "verbose": 0,
+                "epochs": 1000,
+                "batch_size": 32,
+                "callbacks": [
```
It is potentially cleaner to include by default the DEC callback, and check (for example given the repr dunder) whether the list of custom callbacks passed by the user includes DEC, and add it otherwise.

ampudia19 commented on trieste/models/keras/models.py:
```diff
@@ -334,25 +343,159 @@ def optimize(self, dataset: Dataset) -> None:
+
+        mu, sigma = self.sample_normal_parameters(gamma, v, alpha, beta, num_samples)
+       
+        observation_dist = tfp.distributions.Normal(mu, sigma**0.5)
```
Just to confirm, this observation_dist is a N-dimensional multivariate gaussian, just like a GP, and your sigma are the diagonal elements of the covar matrix.

clinton0313 replied:
Well I guess it technically is that, but really this function instantiates n different normal distributions for each of n data points. mu, and sigma would be nx1 vectors.

ampudia19 commented on trieste/models/keras/architectures.py:
```diff
@@ -233,3 +235,113 @@ def connect_layers(self) -> tuple[tf.Tensor, tf.Tensor]:
+        # be positive. 
+        alpha = self._get_evidence(alpha) + 1
+        output = tf.concat([gamma, v, alpha, beta], axis=-1)
+        return output, output
```
Docstrings should mention this, as it may otherwise be confusing and hard to grasp.

ampudia19 reviewed (commented):
All tests run for me, and the comments I left are clarifications or very minor adjustments. Overall, a very clean implementation!

In regards to trying to further approximate Amini's original results, I have tried to create shallower models that may overfit less on the populated space, but never got his level of smoothness. Seeing the output of your attempts at adding dropout and batch_normalization, the latter seems to cause some trouble for predictions around zero so I'd ignore it. Using dropout does not seem to harm, and you could push it as well.

ampudia19 commented:
Once this is pushed, we should try to merge in the code changes from hrvoje's de_sampler branch

ampudia19 committed:
minor typos

clinton0313 committed:
Added small model fixes and architectures fixes including defaults.

Added a better check for model callback to account for tuples and added appropriate test.

Added model docstrings

Updated deep_evidential notebook but still has development testing code commented out at bottom.

Marked deep_ensemble tests. And small changes to notebook comments.

Removed unecessary utils and old functions. Updated tests to reflect that

clinton0313 committed:
Added docstring for DeepEvidentialCallback

Amended model docstring

Amended tests to reflect changes in default args

clinton0313 committed:
Removed deep evidential trajectory sampler that was not working. Sampler has been moved to branch clinton/der_sampler for future development.

clinton0313 committed:
Type hint change to DeepEvidentialCallback

clinton0313 committed:
Deleted notebook figures and moved them to google drive: https://drive.google.com/drive/folders/0AOVQqJ9_N2blUk9PVA

clinton0313 committed:
Removed DER sampler import

clinton0313 committed:
Updated notebook.

clinton0313 committed:
Adjusted der_model to have predict_log and predict_aleatoric attributes

clinton0313 committed:
Adjusted tests for new features
